### DISCUSSION TITLE
Taskbar 10: reordering items in centered taskbar shifts all items to the right

### DISCUSSION THREAD

**Discussion by Poopooracoocoo** [valinet/ExplorerPatcher#1009] [General]
**Environment:**
- Windows 11 build 10.0.22000.527
ExplorerPatcher version 22000.493.42.30
- I usually use 7+ Taskbar Tweaker, but I exited it and restarted File Explorer before conducting testing
- I have enabled the Windows 10 taskbar. I have switched to the center aligned taskbar. I kept the setting to never combine and show labels. 

**Issues and notes:**
- After initially switching to the center aligned taskbar, the windows in the taskbar are not centered and are slightly to the right.
- Launching a new window, closing a window, or unpinning an app centers all of them.
- Reordering an item in the taskbar shifts all of them to the right. 
- The taskbar items do not automatically recentre when an app is pinned to the taskbar.

I'll be happy to provide more information if needed. Thank you so much for this amazing app and I hope this is clear. :)

**Comment by valinet** [comment]
Hi

Yeah, it's funny actually. I looked at it a bit, I think I know why it happens: I compute the necessary width for the taskbar based on the location of the first and last buttons. But, when you shuffle buttons, in particular the first and the last one, the width becomes smaller, so the whole thing shifts right, because it thinks it needs less width.

Yeah, so in 5 minutes I sketched some complicated workaround, didn't work. Now I have this idea: since when you shuffle items, you have to have the left mouse button pressed, I can refrain from updating the width when the left mouse button is pressed. When dragging taskbar buttons, the total width doesn't actually have any reason to change, as you can only shuffle buttons, not drag them away (to other taskbars, for example, etc).

Tried this, seems to work fine. Only thing, I notice that when moving windows from one monitor to the other, the taskbar doesn't update, since the mouse button is pressed while dragging. Of course. So I changed it to filter (not update the necessary width) **when the mouse button is pressed *and* the focused window is the taskbar**. That should cover it, right?

> After initially switching to the center aligned taskbar, the windows in the taskbar are not centered and are slightly to the right.

Idk, for me it works fine, after the taskbar finishes toggling auto hide, it should refresh itself and everything should be where it is supposed to be.

> Launching a new window, closing a window, or unpinning an app centers all of them.

This is good.

> Reordering an item in the taskbar shifts all of them to the right.

This should be fixed with the patch I mentioned above.

> The taskbar items do not automatically recentre when an app is pinned to the taskbar.

Again, over here it seems to work, pinning items from the Windows 10 Start menu to the taskbar just fine.

Anyway, before pushing the update, I will test it a bit more I guess. I am trying to think whether I miss any case with the logic above. Here's a version, if you want to test as well:

[ep_setup.zip](https://github.com/valinet/ExplorerPatcher/files/8192230/ep_setup.zip)

Thanks

**Reply by Poopooracoocoo** [reply]
>>The taskbar items do not automatically recentre when an app is pinned to the taskbar.
>
>Again, over here it seems to work, pinning items from the Windows 10 Start menu to the taskbar just fine.

I forgot to mention that I pinned items by dragging and dropping from the Windows 10 start menu, and not by right clicking and pressing pin to taskbar. 

The first thing that comes to mind when I see your workaround is touch support. I've yet to test your experimental version though so I can't say how well it works.

**Reply by valinet** [reply]
Touch support is maybe for another time if I don’t get that for free with the current solution. But I think it will work, idk, can’t really test that.
